modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1113 @@
+  Real len_p1_p = (p - p1).norm();
+  return len_p1_p / full_len;
+}
+
+bool
+MeshCut3DUserObject::isInsideCutPlane(const std::vector<Point> & vertices, Point p) const
+{
+  unsigned int n_node = vertices.size();
+
+  Point normal;
+  Point center;
+
+  findElemNormal(vertices, normal, center);
+
+  bool inside = false;
+  unsigned int counter = 0;
+
+  for (unsigned int i = 0; i < n_node; ++i)
+  {
+    unsigned int iplus1 = (i < n_node - 1 ? i + 1 : 0);
+    Point middle2p = p - 0.5 * (vertices[i] + vertices[iplus1]);
+    const Point side_tang = vertices[iplus1] - vertices[i];
+    Point side_norm = side_tang.cross(normal);
+    Xfem::normalizePoint(middle2p);
+    Xfem::normalizePoint(side_norm);
+    if (middle2p * side_norm <= 0.0)
+      counter += 1;
+  }
+  if (counter == n_node)
+    inside = true;
+  return inside;
+}
+
+void
+MeshCut3DUserObject::findElemNormal(const std::vector<Point> & vertices,
+                                    Point & normal,
+                                    Point & center) const
+{
+  unsigned int n_node = vertices.size();
+
+  center(0) = center(1) = center(2) = 0;
+  normal(0) = normal(1) = normal(2) = 0;
+
+  for (unsigned int i = 0; i < n_node; ++i)
+    center += vertices[i];
+  center /= n_node;
+
+  for (unsigned int i = 0; i < n_node; ++i)
+  {
+    unsigned int iplus1(i < n_node - 1 ? i + 1 : 0);
+    Point ray1 = vertices[i] - center;
+    Point ray2 = vertices[iplus1] - center;
+    normal += ray1.cross(ray2);
+  }
+  normal /= n_node;
+  Xfem::normalizePoint(normal);
+}
+
+void
+MeshCut3DUserObject::findBoundaryNodes()
+{
+  _boundary_map.clear();
+
+  unsigned int n_nodes = _cut_mesh->n_nodes();
+  std::vector<Real> angle(n_nodes, 0);
+  std::vector<dof_id_type> node_id(_cut_elem_nnode);
+
+  const auto elem_begin = _cut_mesh->elements_begin();
+  auto elem_end = _cut_mesh->elements_end();
+  auto elem_it = elem_begin;
+
+  for (; elem_it != elem_end; ++elem_it)
+  {
+    Elem * cut_elem = *elem_it;
+
+    std::vector<std::vector<Real>> vertices(_cut_elem_nnode, std::vector<Real>(LIBMESH_DIM));
+
+    for (unsigned int i = 0; i < _cut_elem_nnode; ++i)
+    {
+      Node * this_node = cut_elem->get_node(i);
+      Point & this_point = *this_node;
+      for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
+        vertices[i][j] = this_point(j);
+      node_id[i] = this_node->id();
+    }
+
+    for (unsigned int i = 0; i < 3; ++i)
+      mooseAssert(node_id[i] < n_nodes, "Node ID is out of range");
+    angle[node_id[0]] += Xfem::angle_rad_3d(&vertices[2][0], &vertices[0][0], &vertices[1][0]);
+    angle[node_id[1]] += Xfem::angle_rad_3d(&vertices[0][0], &vertices[1][0], &vertices[2][0]);
+    angle[node_id[2]] += Xfem::angle_rad_3d(&vertices[1][0], &vertices[2][0], &vertices[0][0]);
+  }
+
+  for (unsigned int i = 0; i < n_nodes; ++i)
+  {
+    if (!MooseUtils::relativeFuzzyEqual(angle[i], libMesh::pi * 2))
+    {
+      auto ptr = std::make_shared<CutPoint>();
+      _boundary_map[i] = ptr;
```

Comment:
`_boundary_map` would be better as a container for the CutPoint objects (`std::map<CutPoint>`) rather than a container of shared pointers. Also, I think you should use `this_node->id()` as the key in the map, rather than `i`.

Reply:
I got rid of the definition of CutPoint, and simplified the code here.  It is better now.